### PR TITLE
Add support for tracking static DHCP hostnames separately

### PR DESCRIPTION
**Pull request by DroidFreak32**
* This will let Pi-Hole resolve hostnames even if they are not active.
* Addresses #3600

Signed-off-by: DroidFreak32 <rushabshah32@gmail.com>

---
- **What does this PR aim to accomplish?:**

This PR will add support to have a separate file that tracks DHCP hostnames
With this approach, any Static leases can still be resolved even if they are not under active leases.
We no longer will have to maintain a separate Local DNS entry in situations where one needs to constantly update DHCP static leases and has PiHole as their primary DNS resolver.

- **How does this PR accomplish the above?:**

This PR creates 2 separate functions to add and remove entries from a new custom DNS list file `/etc/pihole/static-hostnames.list`
These functions are then called whenever a user adds/removes a static lease in the GUI (Related PR in AdminLTE repo)
Since DHCP lease functions in the WebUI use MAC address as the key, the Add function of this PR will write 2 lines to the list file with the format
```
#<mac address>
<IP address> <hostname>
```
The remove function of this PR will only delete the line right after the matching MAC.
This is done to have minimal changes to the existing code & functions and avoid potential regressions elsewhere

- **What documentation changes (if any) are needed to support this PR?:**

*Replace this with a detailed list of any necessary changes*

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
- [x] I have read the above and my PR is ready for review. _Check this box to confirm_

### CODE REVIEW

**Comment by DroidFreak32** [comment]
Force-push was to fix shellcheck failure

File: advanced/Scripts/webpage.sh [hunk 1]
```diff
@@ -568,6 +569,52 @@ RemoveDHCPStaticAddress() {
...
+    mac="${args[2]}"
+    ip="${args[3]}"
+    host="${args[4]}"
+    reload="${args[5]}"
+
+    validHost="$(checkDomain "${host}")"
+    if [[ -n "${validHost}" ]]; then
+        if valid_ip "${ip}" || valid_ip6 "${ip}" ; then
+            # Mac address will be used as a key to support RemoveDHCPStaticHostname
+            printf "#%s\n%s %s\n" "${mac}" "${ip}" "${validHost}" >> "${dhcpstatichostnames}"
```
Comment (yubiuser) [inline advanced/Scripts/webpage.sh:584]: Did you check if this is actually working? I'm not sure if `dnsmasq` accepts ab `addn-hosts` file with a syntax other than `ip hostname`.

Comment (yubiuser) [reply advanced/Scripts/webpage.sh:584]: Also consider `addKey()` from `utils.sh`

Comment (DroidFreak32) [reply advanced/Scripts/webpage.sh:584]: >  I'm not sure if dnsmasq accepts ab addn-hosts file with a syntax other than ip hostname.  
Yes, the format dnsmasq uses is the same as entries in [/etc/hosts](https://linux.die.net/man/8/dnsmasq)
I tested this by cherry picking this change on the master branch
Addition/romoval works as expected. Existing entries in static leases are unaffected by this change.
> Also consider addKey() from utils.sh  
This suggestion is for adding the new addn-hosts, correct? I will test & update the PR then

Comment (yubiuser) [reply advanced/Scripts/webpage.sh:584]: > Yes, the format dnsmasq uses is the same as entries in [/etc/hosts](https://linux.die.net/man/8/dnsmasq)
I tested this by cherry picking this change on the master branch
Addition/removal works as expected. Existing entries in static leases are unaffected by this change.

It's not working if you add the `MAC`.

```
Aug  8 17:00:39 dnsmasq[14014]: bad address at /etc/pihole/custom.list line 17
Aug  8 17:00:39 dnsmasq[14014]: read /etc/pihole/custom.list - 16 addresses

rockpi@rockpi-4b:~$ wc -l /etc/pihole/custom.list 
17 /etc/pihole/custom.list

rockpi@rockpi-4b:~$ sed -n 17p /etc/pihole/custom.list 
80:86:F2:9C:57:16 1.2.3.4 test
```

Comment (DroidFreak32) [reply advanced/Scripts/webpage.sh:585]: @yubiuser did you add this entry manually to this file?
The entry should look like:
```
#80:86:F2:9C:57:16
1.2.3.4 test
```

Comment (yubiuser) [reply advanced/Scripts/webpage.sh:585]: Sorry, sure. I missed the `#`

**Comment by yubiuser** [comment]
See: https://github.com/pi-hole/AdminLTE/pull/2289#issuecomment-1217817084

**Pull request closed without merging**